src/Data/Text/Encoding.hs

```diff
@@ -187,32 +189,295 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
 foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
     :: Ptr Word8 -> Ptr Word8 -> IO CSize
 
-isValidBS :: ByteString -> Bool
+data Progression
+  = WriteAndAdvance Char Int
+  | NeedMore
+  | Invalid
+
+decodeChunks ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  (Bits w, Num w, Storable w)
+  => w
+  -> ( Int
+    -> ByteString
+    -> A.MArray s
+    -> Int
+    -> Int
+    -> Int
+    -> ST s (Maybe (
+        (A.MArray s -> Int -> Int -> Int -> ST s (DecodeResult Text ByteString w))
+        -> ST s (DecodeResult Text ByteString w))
+      )
+    )
+  -> ((Int -> Word8) -> Int -> Int -> Progression)
+  -> ByteString
+  -> ByteString
+  -> ST s (DecodeResult Text ByteString w)
+decodeChunks w bulkDecodeF incrementalDecodeF bs1@(B.length -> len1) bs2@(B.length -> len2) = do
+  marr <- A.new len'
+  outer marr len' 0 0
+  where
+    wordByteSize = sizeOf w
+
+    index :: Int -> Word8
+    index i
+      | i < len1  = B.index bs1 i
+      | otherwise = B.index bs2 $ i - len1
+
+    len :: Int
+    len = len1 + len2
+    len' :: Int
+    len' = (len `div` wordByteSize) + 4
+
+    outer dst dstLen = inner
+      where
+        inner srcOff dstOff
+          | srcOff < len1 = decode 0 bs1
+          | srcOff < len = decode len1 bs2
+          -- finished (for now)
+          | otherwise = goodSoFar
+          where
+            decode bsOff bs = do
+              mOuterArgs <- bulkDecodeF bsOff bs dst dstLen srcOff dstOff
```

Comment:
Could this be O(n^(2)) in the case where `bs` is a bytestring which is almost UTF-8 except for an invalid byte at the end? I think what happens then is that `bulkDecodeF` scans the string, fails at the end, then `incrementalDecodeF` is called to advance a few bytes, and loop again, with `bulkDecodeF` being called at every iteration.

Reply:
Instead of `bulkDecodeF`, `decodeChunks` could accept a function that counts the number of continuous `Word8`s where the encodings overlap between UTF-8 and the encoding in question (e.g. `c_is_ascii`). Once it knows the length, it can copy that stretch of the bytestring, and then switch to `incrementalDecodeF` for the next code point. This also would require a different function for more than 8-bit word encodings, and simdutf might have to be abandoned.

Reply:
That sounds like a good plan.